### PR TITLE
chore: remove `type: module`, create `openmct-e2e` subpackage

### LINKED ISSUE #7602: Require/import of `openmct` doesn't work as expected
#### Summary

see https://github.com/nasa/openmct/pull/7590 for details

### PR DESCRIPTION
**Pull request by ozyx**
Closes #7602 

### Describe your changes:

Now this is a story all about how our project got flipped, turned upside down--

1iPe0G

**TL;DR:** We took `openmct` to `type: 'module'` too soon.

---

A few months back, we [converted the entire project to using ESM import/export syntax](https://github.com/nasa/openmct/pull/7331). We switched the package itself to `type: module` because running unit/e2e tests were failing (as we couldn't use `import` or `export` keywords unless the package was of `type: module` or the file had an `*.mjs` extension).

Even though our package was of `type: module`, under the hood Webpack was compiling our source using a UMD wrapper which is meant to support imports in both node and browser environments. However, this never worked properly because we set the package type as 'module' (when it really wasn't!). Consumers would attempt to `import` the package only to find that it defined no exports. They would then try to `require` the package, only to be denied under the basis that you cannot `require` an ESModule. In fact, [webpack currently does not support packaging a library as an ESModule](https://github.com/webpack/webpack/issues/2933). As it stands, webpack is a dead-end for `openmct` as a true ESModule.

This PR fixes a ton of misconfiguration with webpack, switches our package type off of `module`, and specifies paths and entry points for both `import` and `require` syntax.

In addition, and to avoid having to change the extensions of all e2e test files to *.mjs, this PR initializes a new npm sub-package `openmct-e2e` of `type: module`. This package is specified as a `workspace` of `openmct`, so a single `node_modules` folder is shared between the two projects (same as it ever was).

See the corresponding PRs for [`openmct-yamcs`](https://github.com/akhenry/openmct-yamcs/pull/446) and `viper-openmct`.

### Brief Summary of Changes

- Removes `e2e` folder from our distributable
- Removes `type: module` from `package.json`
- Initializes a new package `openmct-e2e` which exports our e2e testing framework
- Adds `e2e` folder as a workspace to `openmct` so they share `node_modules`
- `webpack.*.js` -> `webpack.*.mjs`
- Specifies `module` and `exports` in `package.json`
- Updates npm e2e test scripts to use `--workspace` to call scripts in e2e folder
- Modifies `.tsconfig.json` settings to generate TypeScript types in `dist/types/index.d.ts` as opposed to creating individual files.

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this a [notable change](../docs/src/process/release.md) that will require a special callout in the release notes? For example, will this break compatibility with existing APIs or projects that consume these plugins?

### Author Checklist

* [x] Changes address original issue?
* [x] Tests included and/or updated with changes?
* [x] Has this been smoke tested?
* [x] Have you associated this PR with a `type:` label? Note: this is not necessarily the same as the original issue.
* [x] Have you associated a milestone with this PR? Note: leave blank if unsure.
* [x] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [x] Changes appear to address issue?
* [x] Reviewer has tested changes by following the provided instructions?
* [ ] Changes appear not to be breaking changes?
* [x] Appropriate automated tests included?
* [x] Code style and in-line documentation are appropriate?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -2,8 +2,12 @@
   "name": "openmct",
   "version": "4.0.0-next",
   "description": "The Open MCT core platform",
-  "type": "module",
+  "module": "dist/openmct.js",
   "main": "dist/openmct.js",
+  "exports": {
+    "import": "./dist/openmct.js",
+    "require": "./dist/openmct.js"
```
Comment (shefalijoshi) [inline package.json:9]: Aaaah!

**Comment by shefalijoshi** [comment]
@ozyx This is perfect! I tried it with the nirvss project and it now works really well. Thank you!

File: package.json [hunk 2]
```diff
@@ -91,39 +96,39 @@
...
-    "test:e2e:updatesnapshots": "npx playwright test --config=e2e/playwright-ci.config.js --project=chrome --grep @snapshot --update-snapshots",
-    "test:e2e:visual:ci": "percy exec --config ./e2e/.percy.ci.yml --partial -- npx playwright test --config=e2e/playwright-visual-a11y.config.js --project=chrome --grep-invert @unstable",
-    "test:e2e:visual:full": "percy exec --config ./e2e/.percy.nightly.yml -- npx playwright test --config=e2e/playwright-visual-a11y.config.js --grep-invert @unstable",
-    "test:e2e:full": "npx playwright test --config=e2e/playwright-ci.config.js --grep-invert @couchdb",
-    "test:e2e:watch": "npx playwright test --ui --config=e2e/playwright-watch.config.js",
-    "test:perf:contract": "npx playwright test --config=e2e/playwright-performance-dev.config.js",
-    "test:perf:localhost": "npx playwright test --config=e2e/playwright-performance-prod.config.js --project=chrome",
-    "test:perf:memory": "npx playwright test --config=e2e/playwright-performance-prod.config.js --project=chrome-memory",
+    "test:e2e": "npm test --workspace e2e",
```
Comment (scottbell) [inline package.json:115]: much better

**Review by scottbell — APPROVED** [review]
Nice work! We'll need to let others know who use Open MCT that this will require them updating how they import us. I tested core, and using the accompanying PR, and everything appeared to work well.

File: e2e/README.md [hunk 3]
```diff
@@ -205,7 +205,7 @@ Our file structure follows the type of type of testing being excercised at the e
 |`./appActions.js`             | Contains common methods which can be leveraged by test case authors to quickly move through the application when writing new tests.|
 |`./baseFixture.js`            | Contains base fixtures which only extend default `@playwright/test` functionality. The expectation is that these fixtures will be removed as the native Playwright API improves|
 
-Our functional tests end in `*.e2e.spec.js`, visual tests in `*.visual.spec.js` and performance tests in `*.perf.spec.js`.
+Our functional tests end in `*.e2e.spec.js`, visual tests in `*.visual.spec.js` and performance tests in `*.perf.spec.mjs`.
```
Comment (akhenry) [inline e2e/README.md:208]: Do the other test types need updated here as well?

File: e2e/test-data/overlay_plot_storage.json [hunk 4]
```diff
@@ -6,11 +6,11 @@
       "localStorage": [
```
Comment (akhenry) [inline e2e/test-data/overlay_plot_storage.json:6]: What's the change here, formatting?

File: e2e/tests/functional/planning/ganttChart.e2e.spec.mjs [hunk 5]
```diff
@@ -132,3 +131,58 @@ test.describe('Gantt Chart', () => {
     );
   });
 });
+
```
Comment (akhenry) [inline e2e/tests/functional/planning/ganttChart.e2e.spec.mjs:134]: These changes look unrelated to the rest. This a resolved merge conflict?

File: e2e/tests/functional/plugins/clocks/remoteClock.e2e.spec.mjs [hunk 6]
```diff
@@ -22,7 +22,7 @@
 
 // FIXME: Remove this eslint exception once tests are implemented
 // eslint-disable-next-line no-unused-vars
-import { expect, test } from '../../../../baseFixtures.js';
+import { expect, test } from '../../../../baseFixtures.mjs';
```
[the remark this answers is not in the document]
Comment (akhenry) [reply e2e/tests/functional/plugins/clocks/remoteClock.e2e.spec.mjs:25]: Let's fix this before merge.

File: e2e/tests/functional/plugins/importAndExportAsJSON/exportAsJson.e2e.spec.mjs [hunk 7]
```diff
@@ -185,7 +185,7 @@ test.describe('ExportAsJSON ProgressBar @couchdb', () => {
 });
 
 /**
- * Retrieves the first key from the 'openmct' property of the provided JSON object.
+ * Retrieves the first key from the 'openmct' property of the provided mjsON object.
```
Comment (akhenry) [inline e2e/tests/functional/plugins/importAndExportAsJSON/exportAsJson.e2e.spec.mjs:188]: find/replace error.

File: e2e/tests/functional/plugins/importAndExportAsJSON/importAsJson.e2e.spec.mjs [hunk 8]
```diff
@@ -26,7 +26,7 @@
 
 // FIXME: Remove this eslint exception once tests are implemented
 // eslint-disable-next-line no-unused-vars
-import { expect, test } from '../../../../baseFixtures.js';
+import { expect, test } from '../../../../baseFixtures.mjs';
```
[the remark this answers is not in the document]
Comment (akhenry) [reply e2e/tests/functional/plugins/importAndExportAsJSON/importAsJson.e2e.spec.mjs:29]: Let's fix this as well.

File: openmct.js [hunk 9]
```diff
@@ -20,15 +20,6 @@
  * at runtime from the About dialog for additional information.
  *****************************************************************************/
 
-const matcher = /\/openmct.js$/;
```
Comment (akhenry) [inline openmct.js:23]: 👍

File: src/plugins/notebook/utils/painterroInstance.js [hunk 10]
```diff
@@ -46,7 +46,7 @@ export default class PainterroInstance {
     this.config.id = this.elementId;
     this.config.saveHandler = this.saveHandler.bind(this);
 
-    this.painterro = Painterro.default(this.config);
```
Comment (akhenry) [inline src/plugins/notebook/utils/painterroInstance.js:49]: Nice.

File: package.json (hunk 1 above)
Comment (akhenry) [reply package.json:10]: nice 👍

**Review by akhenry — DISMISSED** [review]
Changes look great. Looks like there are some new conflicts due to all the recent merging, and two github warnings we should address before merge. Thanks!

File: e2e/README.md (hunk 3 above)
Comment (ozyx) [reply e2e/README.md:208]: Ooh good catch, this is actually a typo

File: e2e/test-data/overlay_plot_storage.json (hunk 4 above)
Comment (ozyx) [reply e2e/test-data/overlay_plot_storage.json:6]: Yeah just a lint

File: e2e/index.js [hunk 11]
```diff
@@ -0,0 +1,8 @@
+// Import everything from the specific fixture files
+import * as appActions from './appActions.js';
+import * as avpFixtures from './avpFixtures.js';
+import * as baseFixtures from './baseFixtures.js';
+import * as pluginFixtures from './pluginFixtures.js';
+
+// Export these as named exports
+export { appActions, avpFixtures, baseFixtures, pluginFixtures };
```
Comment (unlikelyzero) [inline e2e/index.js:8]: we don't want folks using avpFixture in other repos. Just in vanilla openmct

File: e2e/package.json [hunk 12]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "openmct-e2e",
+  "version": "4.0.0-next",
```
Comment (unlikelyzero) [inline e2e/package.json:3]: this should always match `openmct's package.json`. Is there an easy way to manage that based on monorepo tooling?

File: e2e/playwright-ci.config.js [hunk 13]
```diff
@@ -15,6 +15,7 @@ const config = {
   timeout: 60 * 1000,
   webServer: {
```
Comment (unlikelyzero) [inline e2e/playwright-ci.config.js:16]: we should ignore this file when publishing to npm

File: e2e/playwright-local.config.js [hunk 14]
```diff
@@ -10,6 +10,7 @@ const config = {
   timeout: 30 * 1000,
   webServer: {
```
Comment (unlikelyzero) [inline e2e/playwright-local.config.js:11]: same

File: e2e/playwright-mobile.config.js [hunk 15]
```diff
@@ -14,6 +14,7 @@ const config = {
   timeout: 30 * 1000,
   webServer: {
     command: 'npm run start:coverage',
```
Comment (unlikelyzero) [inline e2e/playwright-mobile.config.js:16]: here

**Review by unlikelyzero — CHANGES_REQUESTED** [review]
Two small changes:
1. We should not be exposing all avpFixtures in other projects. viper should never attempt to publish a percy check!
2. We need to .npmignore our e2e config files

**Review by unlikelyzero — APPROVED** [review]

**Comment by unlikelyzero** [comment]
@ozyx just needs to add `npmignore` for the config files. `avpFixtures` is not a problem

File: packages/e2e/.npmignore [hunk 16]
```diff
@@ -0,0 +1,7 @@
+*
+!appActions.js
```
Comment (unlikelyzero) [inline packages/e2e/.npmignore:2]: I think we're going to want to include the test s but not right now

**Review by unlikelyzero — APPROVED** [review]
